crates/runtime/src/memfd.rs

```diff
@@ -0,0 +1,566 @@
+        // last use of this slot left it larger. This also lets us
+        // skip an mmap the first time a MemFdSlot is used, because we
+        // require the caller to give us a fixed address in an
+        // already-mmaped-with-anon-memory region. This is important
+        // for the on-demand allocator.
+        if self.image.is_some() {
+            self.map_anon_memory(rustix::io::ProtFlags::empty())
+                .map_err(|e| InstantiationError::Resource(e.into()))?;
+        } else if initial_size_bytes < self.initial_size {
+            self.set_protection(
+                initial_size_bytes,
+                self.initial_size,
+                rustix::io::MprotectFlags::empty(),
+            )
+            .map_err(|e| InstantiationError::Resource(e.into()))?;
+        }
+
+        // The initial memory image, if given. If not, we just get a
+        // memory filled with zeroes.
+        if let Some(image) = maybe_image {
+            if image.len > 0 {
+                let image = image.clone();
+
+                unsafe {
+                    let ptr = rustix::io::mmap(
+                        (self.base + image.offset) as *mut c_void,
+                        image.len,
+                        rustix::io::ProtFlags::READ | rustix::io::ProtFlags::WRITE,
+                        rustix::io::MapFlags::PRIVATE | rustix::io::MapFlags::FIXED,
+                        image.fd.as_file(),
+                        image.offset as u64,
+                    )
+                    .map_err(|e| InstantiationError::Resource(e.into()))?;
+                    assert_eq!(ptr as usize, self.base + image.offset);
+                }
+
+                self.image = Some(image);
+            }
+        }
+
+        // mprotect the initial `initial_size_bytes` to be accessible.
+        self.initial_size = initial_size_bytes;
+        self.cur_size = initial_size_bytes;
+        self.set_protection(
+            0,
+            initial_size_bytes,
+            rustix::io::MprotectFlags::READ | rustix::io::MprotectFlags::WRITE,
+        )
+        .map_err(|e| InstantiationError::Resource(e.into()))?;
+
+        self.dirty = true;
+        Ok(())
+    }
+
+    pub(crate) fn clear_and_remain_ready(&mut self) -> Result<()> {
+        assert!(self.dirty);
+        // madvise the image range. This will throw away dirty pages,
+        // which are CoW-private pages on top of the initial heap
+        // image memfd.
+        unsafe {
+            rustix::io::madvise(
+                self.base as *mut c_void,
+                self.static_size,
+                rustix::io::Advice::LinuxDontNeed,
+            )?;
+        }
+
+        // mprotect the initial heap region beyond the initial heap size back to PROT_NONE.
+        self.set_protection(
+            self.initial_size,
+            self.static_size - self.initial_size,
+            rustix::io::MprotectFlags::empty(),
+        )?;
```

Comment:
One thought I had reading over this, is this strictly necessary? For example if we didn't track `initial_size` and instead only tracked `cur_size` then we could leave the entire mapping as zero here and on reuse we'd hit the `if initial_size_bytes < self.initial_size` case (although afterwards it'd be `self.cur_size`).

I don't think that'd actually result in any fewer syscalls happening in any cases though.

Reply:
Oh actually this may not work out, I can see how a grown heap would then, on resue, re-`mmap` the CoW overlay.